llvm/include/llvm/IR/RepoDefinition.h

```diff
@@ -107,56 +103,116 @@ struct GODigestState {
+  /// \param GO The global object whose digest is to be computed.
+  /// \return The global object's digest value.
+  DigestType calculateDigestUseRepoDefinition(const GlobalObject *GO);
+
+  bool isChanged() const { return Changed; }
+
+  const GOInfoMap &getGOInfoMap() const { return GOIMap; }
+
+  const MemoizedHashes &getGOHashCache() const { return GOHashCache; }
+
+private:
+  /// Get global object hash value from memoized hashes.
+  /// \param GO The global object.
+  /// \return A tuple of global object's hash value and a bool which is true if
+  /// GO's hash is cached.
+  std::tuple<DigestType, bool> getGOHash(const GlobalObject *GO) const {
+    const auto Iter = GOHashCache.find_as(GO);
+    return (Iter != GOHashCache.end()) ? std::make_tuple(Iter->second, true)
+                                       : std::make_tuple(NullDigest, false);
   }
-  // Update G's dependencies.
-  GOInfo &GInfo = GOI[G];
-  GInfo.InitialDigest = std::move(Result.InitialDigest);
-  GInfo.Dependencies = std::move(Result.Dependencies);
-}
-
-// Create a global object information map and calculate the number of hashed
-// functions and variable  inside of the Module M.
-/// \param M Called module.
-/// \returns a pair containing a global object information map and a structure
-/// holding the number of hashed global variables and functions.
-std::pair<GOInfoMap, GONumber> calculateGONumAndGOIMap(Module &M);
+
+  /// Construct the global object information map (GOIMap) and calculate the
+  /// number of hashed functions and variables inside of the Module M.
+  void calculateGONumAndGOIMap(const Module &M);
+
+  /// Compute the hash value for the given global object GO.
+  /// \param GO The global object.
+  /// \param HasRepoDefinitionMD true if using the digest of GO's dependencies
+  ///        which is stored in the RepoDefinition metadata.
+  /// \return The global object's digest value.
+  DigestType calculateDigest(const GlobalObject *GO, bool HasRepoDefinitionMD);
+
+  /// Calculate the initial hash value, dependencies and contributions for the
+  /// global object 'G' and store these informaion in the GOImap.
+  /// \param G Calculated global object.
+  template <typename GlobalType> void calculateGOInfo(const GlobalType *G);
+  void calculateGOInfo(const GlobalObject *GO);
+
+  /// Accumulate the GO's digest and get its GODigestState.
+  /// \param GO Calculated global object.
+  /// \param GOHash A object that is used to calculate the GO's hash value.
+  /// \param HasRepoDefinitionMD true if using the digest of GO's dependencies
+  ///        which is stored in the RepoDefinition metadata.
+  /// \returns a structure containing the GO's state.
+  GODigestState accumulateGODigest(const GlobalObject *GO, MD5 &GOHash,
+                                   bool HasRepoDefinitionMD);
+
+  /// Loop through the contributions and add the initial digest of each global
+  /// object inside of the contributions to GOHash.
```

Comment:
I think that "Loop through" and "... of each global object" repeat each other.

How about:
"Add the initial digest of each contribution global object to GOHash."

Reply:
Much better, thanks.